rfc/2/index.md

```diff
@@ -0,0 +1,459 @@
+This is a fairly cheap operation, because only json files are touched.
+Zarr version 2 and 3 metadata could even live side-by-side in the same hierarchy.
+There are [scripts available](https://github.com/scalableminds/zarrita/blob/8155761/zarrita/array_v2.py#L452-L559) that can migrate the metadata automatically.
+
+### Notable changes in Zarr v3
+
+There are a few notable changes that Zarr v3 brings for OME-Zarr:
+
+- Array and group metadata including attributes are now stored in `zarr.json` files instead of `.zarray`, `.zgroup` and `.zattrs`. The attributes are now represented in an `attributes` key within the `zarr.json` files.
+- Arrays specify a `chunk_key_encoding` that controls under what naming scheme chunks are stored. This is similar to the previous `dimension_separator` attribute. As part of this proposal, OME-Zarr will support all valid chunk key encodings instead of mandating a `/` dimension separator.
+- There is a new codec pipeline concept that unifies filters and compression codecs as well as array-to-byte serialization including endianness and index ordering configuration.
+  OME-Zarr will support all codecs in the specification.
+  In the future there will likely be additional codecs including image-specific codecs that OME-Zarr would automatically adopt.
+  This is the current list of available codecs:
+  - [blosc](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/blosc/v1.0.html) for compression
+  - [gzip](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/gzip/v1.0.html) for compression
+  - [transpose](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/transpose/v1.0.html) for transposing the data before serialization, e.g. to support C and F orders
+  - [bytes](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/bytes/v1.0.html) for serializing arrays to byte streams with configurable endianness
+  - [crc32c](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/crc32c/v1.0.html) for decorating chunks with a checksum
+  - [sharding_indexed](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/sharding-indexed/v1.0.html) for storing sharded arrays (see below).
+
+The Zarr specification does not prescribe the support stores for Zarr hierarchies.
+HTTP(S), File system, S3, GCS, and Zip files are commonly used stores.
+
+With this proposal all features of the Zarr specification are allowed in OME-Zarr.
+In the future, the OME-Zarr community MAY decide to restrict the allowed feature set.
+
+### Changes to the OME-Zarr metadata
+
+While the adoption of Zarr v3 does not strictly require changes to the OME-Zarr metadata, this proposal contains changes to align with community conventions and ease implementation:
+
+- OME-Zarr metadata will be stored under a dedicated key in the Zarr array or group attributes. The key will be a well-known URI of the OME-NGFF specification with a version number, e.g. `https://ngff.openmicroscopy.org/0.6`.
```

Comment:
"E.g" or that specific one for ever? Or it would be over starting with that specific uri and version component possibly changing?

Why not just "ngff" and have dedicated and clearly defined field "Version"? That would simplify accessing such key (instead of matching by leading uri if accommodating multiple versions).

Reply:
The idea was to reference the version in the key. Yes, that would require a prefix scan, which I don't think is a big problem because most implementation parse the entire json, anyways. The upside is that you could have multiple OME metadata versions in the same document, e.g. `https://ngff.openmicroscopy.org/0.6` and `https://ngff.openmicroscopy.org/1.0`.
@joshmoore suggested this approach in favor of having a generic "ome" key with an additional "version" key in https://github.com/ome/ngff/pull/206#issuecomment-1746870625

Reply:
That's "cute" -- thanks for the explanation/reference!  I wonder if use of a versioned URI to "allocate" a namespace here was ra- by a wider Zarr community?  not that it would not be permitted but rather since I believe it is first/unique use there (IIRC it standed out whenever I saw it among other keys) and wider Zarr might either like to adopt it wider or raise some concerns.

My immediate question is: what should be a "standard" behavior when multiple versions are encountered? could be "the latter supported is taken".  And either such behavior also should be formalized at Zarr or NGFF levels. Might be worth spelling out as well ahead of time if allowing for such multi-version specification.

Reply:
> That's "cute" -- thanks for the explanation/reference! I wonder if use of a versioned URI to "allocate" a namespace here was ra- by a wider Zarr community? not that it would not be permitted but rather since I believe it is first/unique use there (IIRC it standed out whenever I saw it among other keys) and wider Zarr might either like to adopt it wider or raise some concerns.

There is an [open discussion on the Zarr level about metadata conventions](https://zarr.dev/zeps/draft/ZEP0004.html), of which NGFF/OME-Zarr is one. That discussion has not reached a conclusion yet. And, I don't think the OME community necessarily needs to wait for a conclusion.

Generally, URLs are used in other places in the Zarr spec, e.g. for [names of extensions](https://zarr-specs.readthedocs.io/en/latest/v3/core/v3.0.html#extension-points). So, it does not seem entirely foreign.

> My immediate question is: what should be a "standard" behavior when multiple versions are encountered? could be "the latter supported is taken". And either such behavior also should be formalized at Zarr or NGFF levels. Might be worth spelling out as well ahead of time if allowing for such multi-version specification.

That is currently not defined. I almost think that implementations should decide which metadata they should follow. My thinking is that image data and metadata must match. That is true for every version of the metadata that may exist. So, implementations can choose which version they understand and use that to interpret the image data. If implementations support multiple versions, it would be reasonable to use the most recent one.

For example, you could have an OME-Zarr image that has the 0.4 metadata (based on Zarr v2) and the 0.6 metadata (based on Zarr v3). That is possible, because the chunk data is mostly compatible and just the json files differ. Implementations that don't support Zarr v3 can still read the v2 version and vice versa.